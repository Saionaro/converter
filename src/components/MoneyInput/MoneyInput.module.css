.root {
  display: inline-flex;
  align-items: center;
  position: relative;
  cursor: text;
}

.currency {
  order: 1;
  font-size: 32px;
  font-weight: 200;
  transition-property: color;
  transition-duration: var(--time-interaction);
}

.input {
  order: 2;
  color: transparent;
  padding: 12px 6px;
  font-size: 0;
  width: 300px;
  height: 60px;
  border: none;
}

.input:focus {
  outline: none;
}

.view {
  position: absolute;
  text-align: right;
  font-size: 32px;
  width: 100%;
  top: 50%;
  transform: translateY(-50%);
  transition-property: color;
  transition-duration: var(--time-interaction);
}

.input:focus ~ .currency,
.input:focus ~ .view {
  color: var(--color-primary);
}

.inactive {
  color: var(--color-gray-light);
}

.active {
  position: relative;
}

.input:focus ~ .view .active::after {
  content: "";
  height: 100%;
  width: 2px;
  position: absolute;
  top: 0;
  right: 0;
  background-color: var(--color-gray-light);
  animation: blink;
  animation-duration: 1s;
  animation-iteration-count: infinite;
}

.input:disabled {
  background-color: transparent;
}

@keyframes blink {
  0% {
    opacity: 0;
  }
  50% {
    opacity: 1;
  }
  100% {
    opacity: 0;
  }
}
