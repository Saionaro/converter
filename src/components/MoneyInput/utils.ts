const AFTER_DOT_LIMIT = 2;

export const allowedSymbos = {
  "0": true,
  "1": true,
  "2": true,
  "3": true,
  "4": true,
  "5": true,
  "6": true,
  "7": true,
  "8": true,
  "9": true,
} as const;
export const DOT = ".";
export const BACKSPACE = "Backspace";

export function toExport(value: string): string {
  let cleanRes = "";
  let afterDotCount = 0;
  let hasDot = false;

  for (const symbol of value) {
    if (symbol in allowedSymbos) {
      cleanRes += symbol;
    }
    if (symbol === DOT && !hasDot) {
      cleanRes += symbol;
      hasDot = true;
    }
    if (hasDot) afterDotCount++;
    if (afterDotCount > AFTER_DOT_LIMIT) break;
  }

  return cleanRes;
}
