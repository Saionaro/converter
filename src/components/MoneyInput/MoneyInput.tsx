import { useCallback, KeyboardEvent, useRef } from "react";
import cn from "classnames";
import { Currency } from "src/common-types";
import { formatters } from "src/utils/format";

import { toExport, allowedSymbos, BACKSPACE, DOT } from "./utils";
import st from "./MoneyInput.module.css";

interface Props {
  value: string;
  currency: Currency;
  onChange: (val: string) => void;
}

const MAX_LEN = 25;

export function MoneyInput({ value, currency, onChange }: Props) {
  const inputRef = useRef<HTMLInputElement>(null);
  const handleChange = useCallback(
    ({ key }: KeyboardEvent) => {
      let res: null | string = null;
      if (key === BACKSPACE) {
        res = value.slice(0, -1);
      } else if (key in allowedSymbos || key === DOT) {
        res = value + key;
      }
      if (res !== null) {
        onChange(toExport(res));
      }
    },
    [value, onChange]
  );

  return (
    <div className={st.root} onClick={() => inputRef?.current?.focus()}>
      {getDisplay(value, currency)}
      <input
        className={st.input}
        value={value}
        ref={inputRef}
        onKeyUp={handleChange}
        maxLength={MAX_LEN}
        readOnly
      />
    </div>
  );
}

function getDisplay(value: string, currency: Currency) {
  const val = formatters[currency](value);
  const originalParts = value.split(".");
  const formattedParts = val.split(".");
  const decimal = originalParts[1]?.split("") ?? [];
  const formattedDecimal = formattedParts[1].split("");

  const hasDecimal = originalParts.length > 1;

  return (
    <div className={st.view}>
      <span>{formattedParts[0]}</span>
      <span className={cn(!hasDecimal && st.inactive)}>.</span>
      <span className={cn(!decimal[0] && st.inactive)}>
        {formattedDecimal[0]}
      </span>
      <span className={cn(!decimal[1] && st.inactive)}>
        {formattedDecimal[1]}
      </span>
    </div>
  );
}
