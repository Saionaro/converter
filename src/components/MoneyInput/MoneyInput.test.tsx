import React, { ComponentProps } from "react";
import { render, fireEvent, screen, cleanup } from "@testing-library/react";
import { MoneyInput } from "./MoneyInput";

import "@testing-library/jest-dom/extend-expect";

const mockCallback = jest.fn((x) => x);

const renderInput = (params: ComponentProps<typeof MoneyInput>) => {
  params.value = last(mockCallback.mock.results)?.value ?? "";
  cleanup();
  return render(<MoneyInput {...params} />);
};

function last<T>(arr: T[]): T {
  return arr[arr.length - 1];
}

describe("components/MoneyInput", () => {
  test("fit snapshot renders", () => {
    const { container } = renderInput({
      value: "",
      currency: "USD" as const,
      onChange: jest.fn(),
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  test("no extra decimal while typing in input", () => {
    const params = {
      value: "",
      currency: "USD" as const,
      onChange: mockCallback,
    };
    renderInput(params);

    let input = screen.getByTestId("input");
    fireEvent.keyUp(input, { key: "5" });

    expect(mockCallback).toBeCalledTimes(1);
    expect(last(mockCallback.mock.results).value).toBe("5");

    renderInput(params);
    input = screen.getByTestId("input");
    fireEvent.keyUp(input, { key: "5" });

    expect(mockCallback).toBeCalledTimes(2);
    expect(last(mockCallback.mock.results).value).toBe("55");

    renderInput({ ...params });
    input = screen.getByTestId("input");
    fireEvent.keyUp(input, { key: "." });

    expect(mockCallback).toBeCalledTimes(3);
    expect(mockCallback.mock.results[2].value).toBe("55.");

    renderInput(params);
    input = screen.getByTestId("input");
    fireEvent.keyUp(input, { key: "f" });

    expect(mockCallback).toBeCalledTimes(3);
    expect(mockCallback.mock.results[2].value).toBe("55.");

    renderInput(params);
    input = screen.getByTestId("input");
    fireEvent.keyUp(input, { key: "5" });
    renderInput(params);
    input = screen.getByTestId("input");
    fireEvent.keyUp(input, { key: "5" });
    renderInput(params);
    input = screen.getByTestId("input");
    fireEvent.keyUp(input, { key: "5" });

    renderInput({ ...params });

    expect(last(mockCallback.mock.results).value).toBe("55.55");
  });
});
