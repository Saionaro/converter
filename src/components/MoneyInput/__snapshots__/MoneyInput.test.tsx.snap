// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/MoneyInput fit snapshot renders 1`] = `
<div
  role="button"
>
  <input
    data-testid="input"
    inputmode="decimal"
    maxlength="25"
    type="number"
    value=""
  />
  <h2>
    USD
  </h2>
  <div>
    <span
      class=""
    >
      $
    </span>
    <span
      class="undefined"
    >
      0
    </span>
    <span
      class=""
    >
      .
    </span>
    <span
      class=""
    >
      0
    </span>
    <span
      class=""
    >
      0
    </span>
  </div>
</div>
`;
